tests/integration/tui/test_app.py

```diff
@@ -0,0 +1,378 @@
+#!/usr/bin/env python3
+# THIS FILE IS PART OF THE CYLC WORKFLOW ENGINE.
+# Copyright (C) NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import pytest
+import urwid
+
+from cylc.flow.cycling.integer import IntegerPoint
+from cylc.flow.task_state import (
+#     TASK_STATUS_RUNNING,
+    TASK_STATUS_SUCCEEDED,
+#     TASK_STATUS_FAILED,
+#     TASK_STATUS_WAITING,
+)
+from cylc.flow.workflow_status import StopMode
+
+
+def set_task_state(schd, task_states):
+    """Force tasks into the desired states.
+
+    Task states should be of the format (cycle, task, state, is_held).
+    """
+    for cycle, task, state, is_held in task_states:
+        itask = schd.pool.get_task(cycle, task)
+        if not itask:
+            itask = schd.pool.spawn_task(task, cycle, {1})
+        itask.state_reset(state, is_held=is_held)
+        schd.data_store_mgr.delta_task_state(itask)
+        schd.data_store_mgr.increment_graph_window(
+            itask.tokens,
+            cycle,
+            {1},
+        )
+
+
+async def test_tui_basics(raikura):
+    """Test basic Tui interaction with no workflows."""
+    with raikura(size='80,40') as rk:
+        # the app should open
+        rk.compare_screenshot('test-raikura', 'the app should have loaded')
+
+        # "h" should bring up the onscreen help
+        rk.user_input('h')
+        rk.compare_screenshot(
+            'test-raikura-help',
+            'the help screen should be visible'
+        )
+
+        # "q" should close the popup
+        rk.user_input('q')
+        rk.compare_screenshot(
+            'test-raikura',
+            'the help screen should have closed',
+        )
+
+        # "enter" should bring up the context menu
+        rk.user_input('enter')
+        rk.compare_screenshot(
+            'test-raikura-enter',
+            'the context menu should have opened',
+        )
+
+        # "enter" again should close it via the "cancel" button
+        rk.user_input('enter')
+        rk.compare_screenshot(
+            'test-raikura',
+            'the context menu should have closed',
+        )
+
+        # "ctrl d" should exit Tui
```

Comment:
Ought you test `ctrl c` and `ctrl z` cases too?

Reply:
`ctrl z` can't be reasonably tested (you need a genuine interactive terminal).

`ctrl c` doesn't yield the exception.